cpp/src/interop/from_arrow_device.cu

```diff
@@ -0,0 +1,448 @@
+  ArrowSchemaView* schema,
+  const ArrowArray* input,
+  data_type type,
+  bool skip_mask,
+  rmm::cuda_stream_view stream,
+  rmm::mr::device_memory_resource* mr)
+{
+  if (input->length == 0) {
+    return std::make_tuple<column_view, owned_columns_t>(
+      {type,
+       0,
+       nullptr,
+       skip_mask ? nullptr : reinterpret_cast<bitmask_type const*>(input->buffers[0]),
+       0},
+      {});
+  }
+
+  auto offsets_view = column_view{data_type(type_id::INT32),
+                                  static_cast<size_type>(input->offset + input->length) + 1,
+                                  input->buffers[1],
+                                  nullptr,
+                                  0,
+                                  0};
+  return std::make_tuple<column_view, owned_columns_t>(
+    {type,
+     static_cast<size_type>(input->length),
+     input->buffers[2],
+     skip_mask ? nullptr : reinterpret_cast<bitmask_type const*>(input->buffers[0]),
+     static_cast<size_type>(input->null_count),
+     static_cast<size_type>(input->offset),
+     {offsets_view}},
+    {});
+}
+
+template <>
+std::tuple<column_view, owned_columns_t> dispatch_to_cudf_column::operator()<cudf::dictionary32>(
+  ArrowSchemaView* schema,
+  const ArrowArray* input,
+  data_type type,
+  bool skip_mask,
+  rmm::cuda_stream_view stream,
+  rmm::mr::device_memory_resource* mr)
+{
+  ArrowSchemaView keys_schema_view;
+  NANOARROW_THROW_NOT_OK(
+    ArrowSchemaViewInit(&keys_schema_view, schema->schema->dictionary, nullptr));
+
+  auto const keys_type = arrow_to_cudf_type(&keys_schema_view);
+  auto [keys_view, owned_cols] =
+    get_column(&keys_schema_view, input->dictionary, keys_type, true, stream, mr);
+
+  auto const dict_indices_type = [&schema]() -> data_type {
+    switch (schema->storage_type) {
+      case NANOARROW_TYPE_INT8: return data_type(type_id::INT8);
+      case NANOARROW_TYPE_INT16: return data_type(type_id::INT16);
+      case NANOARROW_TYPE_INT32: return data_type(type_id::INT32);
+      case NANOARROW_TYPE_INT64: return data_type(type_id::INT64);
+      case NANOARROW_TYPE_UINT8: return data_type(type_id::UINT8);
+      case NANOARROW_TYPE_UINT16: return data_type(type_id::UINT16);
+      case NANOARROW_TYPE_UINT32: return data_type(type_id::UINT32);
+      case NANOARROW_TYPE_UINT64: return data_type(type_id::UINT64);
+      default: CUDF_FAIL("Unsupported type_id for dictionary indices");
+    }
+  }();
+
+  column_view indices_view = column_view{dict_indices_type,
+                                         static_cast<size_type>(input->offset + input->length),
+                                         input->buffers[1],
+                                         nullptr,
+                                         0,
+                                         0};
+
+  // need to cast the indices to uint32 instead of just using them as-is
+  if (dict_indices_type != data_type{type_id::UINT32}) {
```

Comment:
The developer doc is correct. The dictionary indices can be any unsigned integer type.
The parent type is `DICTIONARY32`, the keys child column is any comparable type (except dictionary) and the indices child is UINT8, UINT16, UINT32, or UINT64.
The 32 in `DICTIONARY32` is a hold over from the original implementation that I'd rather not waste time changing.

Reply:
I was following the existing `from_arrow.cu` implementation which performs a cast if the indices aren't UINT32. If it can be any unsigned integer type, then i'll simplify this and avoid the cast

Reply:
updated this